packages/devtools/lib/src/vm_service_wrapper.dart

```diff
@@ -176,23 +190,90 @@ class VmServiceWrapper implements VmService {
     );
   }
 
-  // TODO(kenzie): keep track of all private methods we are currently using to
-  // share with the VM team and request that they be made public.
-  Future<Response> getCpuProfileTimeline(
-      String isolateId, int origin, int extent) async {
+  @override
+  Future<CpuSamples> getCpuSamples(
+      String isolateId, int timeOriginMicros, int timeExtentMicros) async {
     return _trackFuture(
-        'getCpuProfileTimeline',
-        callMethod(
-          '_getCpuProfileTimeline',
-          isolateId: isolateId,
-          args: {
-            'tags': 'None',
-            'timeOriginMicros': origin,
-            'timeExtentMicros': extent,
-          },
+        'getCpuSamples',
+        _vmService.getCpuSamples(
+          isolateId,
+          timeOriginMicros,
+          timeExtentMicros,
         ));
   }
 
+  Future<CpuProfileData> getCpuProfileTimeline(
+      String isolateId, int origin, int extent) async {
+    if (await isProtocolVersionLessThan(major: 3, minor: 27)) {
+      return _trackFuture(
+          'getCpuProfileTimeline',
+          callMethod(
+            '_getCpuProfileTimeline',
+            isolateId: isolateId,
+            args: {
+              'tags': 'None',
+              'timeOriginMicros': origin,
+              'timeExtentMicros': extent,
+            },
+          )).then((Response response) => CpuProfileData.parse(response.json));
+    }
+    // As of service protocol version 3.27 _getCpuProfileTimeline does not exist
+    // and has been replaced by getCpuSamples. We need to do some processing to
+    // get back to the format we expect.
```

Comment:
Can we add tests for all this processing code?

Reply:
Writing tests for this will probably be pretty involved and will involve either recording or fudging a bunch of samples. If you feel it's worth adding a test or two I can, I just won't be terribly excited about it :-)

Reply:
My only concern is that something could break in this code and our current tests wouldn't catch it.  For these tests, we use stubbed json data in the format we are expecting (which is the format this code should convert sample data into), but our stubbed data becomes irrelevant if something goes wrong in this part of the pipeline.  

Are there already tests for this logic in the VM?

Reply:
That's totally reasonable. This particular logic was never exactly tested in the VM (the trie builder was tested, but not the conversion to the catapult format) so all that's tested explicitly in the VM is the format of the samples.

For more context, the full change is [here](https://dart-review.googlesource.com/c/sdk/+/112265) and you can see what's being tested there. I'd like to land that change sooner than later before but I'm not going to do that until I know DevTools is ready to migrate.

Reply:
I see. I still think it would be wise to have a test for the conversion, either here or in the VM, but I won't block submit over it. Can you add a todo or file an issue to add tests? In the event of something breaking, I can see this being a case that will be difficult to debug or catch without tests.

Reply:
This logic is now tested here: https://dart-review.googlesource.com/c/sdk/+/112265/18/runtime/observatory/tests/service/get_cpu_profile_timeline_rpc_test.dart

Reply:
This logic is now tested in the following VM service test: https://dart-review.googlesource.com/c/sdk/+/112265/18/runtime/observatory/tests/service/get_cpu_profile_timeline_rpc_test.dart

Reply:
great thanks, Ben!